### PR TITLE
feat(raft): add support for joint consensus phase

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This adds support for a joint consensus phase. This feature remains unused for now so the PR is effectively a refactoring that should not result in behavior changes.

### Joint Consensus Support

Writing a configuration entry containing both new and old member lists signals the joint consensus phase. During this phase, the `RaftClusterContext` requires votes and replication from both member sets.

When the configuration entry for the joint consensus phase is committed by a majority of old members and a majority of new members, a new configuration containing only the new member set is written, thus leaving the joint consensus phase. This is usually handled by the leader waiting for the commit notification for the joint consensus configuration entry. In rare cases where the joint consensus configuration entry is committed but the leader fails to write the new configuration, leaving the joint consensus phase it is handled during leader startup when the leader finds a persisted joint consensus configuration.

Currently, we never _enter_ the joint consensus phase because it is only wired up to reconfigure requests through `LeaderRole#onReconfigure` which is unused.

### Serialization and backwards compatibility
We extend the `ConfigurationEntry` to store both old and new member sets. This should be a backwards compatible change because the SBE (de-)serializer detects previous versions and can assume that no old members were set. Similarly, the `Configuration` type also supports old members now. This type is also serialized to and from the `MetaStore` via SBE which again covers the backwards compatibility.
The `ReconfigureRequest` changed to support arbitrary changes to the member set. Usually this would break backwards compatibility (or at least impact rolling updates) but the request type is completely unused so it's okay to break it.

### Testing
This feature required extensive refactoring of `RaftClusterContext` which is a critical component. Unfortunately, this class was only tested implicitly, for example through the raft property tests. This made changes to this class very difficult. This PR adds some initial tests but nowhere close to full coverage.
Before merging, we should let run the randomized raft test for longer than unusual.

relates to #13906

### CODE REVIEW

**Comment by deepthidevaki** [comment]
@oleschoenburg I haven't reviewed the PR yest. But while skipping though it, I see that `ConfigureRequest` is unchanged. It can accept only one set of members. This is the one that Leader sends to the followers to update the in-memory state. As far as I can see it should be updated to support joint consensus. Do you plan to add it in a follow up?

**Comment by oleschoenburg** [comment]
Yes, since changing `ConfigureRequest` is a bit tricky for backwards compatibility I think doing that in a follow up PR would be best.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java [hunk 1]
```diff
@@ -16,79 +16,54 @@
...
 
-  @Override
-  public String toString() {
-    return toStringHelper(this)
-        .add("index", index)
-        .add("time", new TimestampPrinter(time))
-        .add("members", members)
-        .toString();
+  public Set<RaftMember> allMembers() {
+    return new HashSet<>(newMembers) {
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java:63]: :wrench: Double brace initialization is considered an anti-pattern.

File: atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+import io.atomix.raft.impl.RaftContext;
+import io.atomix.raft.storage.system.Configuration;
+import io.atomix.raft.storage.system.MetaStore;
+import java.time.Instant;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
+
+final class RaftClusterContextTest {
```
Comment (deepthidevaki) [inline atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java:29]: :+1: Nice to have tests.

File: atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    context.getMemberContext(new MemberId("3")).setMatchIndex(3);
+    context.getMemberContext(new MemberId("4")).setMatchIndex(4);
+    context.getMemberContext(new MemberId("5")).setMatchIndex(5);
+
+    // then
+    assertThat(context.getQuorumFor(RaftMemberContext::getMatchIndex)).hasValue(4L);
+  }
+
+  @Test
+  void shouldCalculateQuorumDuringJointConsensus() {
```
Comment (deepthidevaki) [inline atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java:215]: :wrench: : Also add test where matchIndex in oldMembers is higher than matchIndex in new members.

File: atomix/cluster/src/test/java/io/atomix/raft/utils/VoteQuorumTest.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+      quorum.succeed(new MemberId("2"));
+      quorum.succeed(new MemberId("4"));
+      quorum.succeed(new MemberId("5"));
+
+      // then
+      verify(callback, only()).accept(true);
+    }
+
+    @Test
+    void shouldWaitWithPartialQuorum() {
```
Comment (deepthidevaki) [inline atomix/cluster/src/test/java/io/atomix/raft/utils/VoteQuorumTest.java:190]: :wrench: Also add tests where oldMembers have quorum and newMembers does not, and vise versa.

**Review by deepthidevaki — APPROVED** [review]
:rocket: Thanks for cleaning up `RaftClusterContext`. Great progress on the topic.

File: atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java [hunk 5]
```diff
@@ -208,72 +164,142 @@ private void createInitialConfig(final Collection<MemberId> cluster) {
...
+              .filter(context -> oldMembers.contains(context.getMember()))
+              .collect(Collectors.toCollection(ArrayList::new));
+      final var newContexts =
+          contexts.stream()
+              .filter(context -> newMembers.contains(context.getMember()))
+              .collect(Collectors.toCollection(ArrayList::new));
+
+      final var oldQuorum = getQuorumFor(oldContexts, calculateMemberValue);
+      final var newQuorum = getQuorumFor(newContexts, calculateMemberValue);
+      if (oldQuorum.isPresent() && newQuorum.isPresent()) {
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java:206]: :wrench: Actually this always returns true. oldQuorum and newQuorum is never empty. Please check it and remove the `else` case if it applies.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java (hunk 1 above)
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java:63]: :disappointed: I just discovered it for the first time. It does look odd though, I'll remove it :+1:

File: atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java [hunk 6]
```diff
@@ -373,6 +381,18 @@ public void configure(final Configuration configuration) {
...
+        remoteMemberContexts.values().stream()
+            .filter(context -> context.getMember().getType() == Type.ACTIVE)
+            .toList();
+
+    replicationTargets =
+        remoteMemberContexts.values().stream()
+            .filter(context -> context.getMember().getType() != Type.INACTIVE)
+            .toList();
+
+    hasRemoteActiveMembers = !activeMembers.isEmpty();
```
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java]: Done in a later commit

**Comment by oleschoenburg** [comment]
I'm running `RandomizedRaftTest` locally:
- [ ] Default tries, 10k operations (the default)
- [ ] Default tries, 20k operations
- [ ] 10x tries, 10k operations
- [ ] 100x tries, 10k operations

File: atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java (hunk 5 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java:206]: :+1: My IDE was showing warning that this is never empty. If this is not true, then no problem.

**Comment by oleschoenburg** [comment]
bors merge

**Comment by oleschoenburg** [comment]
bors merge

**Comment by oleschoenburg** [comment]
bors cancel

This is the second time the unit tests are timing out. I'll need to look into it, better safe than sorry.

**Comment by oleschoenburg** [comment]
Last merge attempt failed because `StreamProcessorTest` did not finish. That seems unrelated. I'm trying to reproduce it locally now.

**Comment by oleschoenburg** [comment]
Yep, `shouldBlockProcessingIfSchedulingBlocks` runs into a deadlock sometimes. This is unrelated so let's try to merge again.

bors merge

**Pull request merged**